In [131]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [132]:
df = pd.read_csv(r"D:\Modules_UCD\COMP47490\Assignment 2\census_22209345.csv", skipinitialspace=True)
df

age workclass  fnlwgt     education  education-num      marital-status  \
0      36   Private  136629  Some-college             10  Married-civ-spouse   
1      56   Private   98361       Masters             14  Married-civ-spouse   
2      32   Private  222548       HS-grad              9            Divorced   
3      24   Private  234640     Assoc-voc             11       Never-married   
4      25   Private   74977  Some-college             10  Married-civ-spouse   
...   ...       ...     ...           ...            ...                 ...   
9995   43   Private  313022       HS-grad              9  Married-civ-spouse   
9996   22   Private  231085  Some-college             10       Never-married   
9997   60         ?  386261     Bachelors             13  Married-spouse-abs   
9998   47   Private  188081       HS-grad              9  Married-civ-spouse   
9999   25   Private  212304       HS-grad              9       Never-married   

             occupation relationship   race     sex  capital-gain  \
0     Machine-op-inspct      Husband  White    Male             0   
1        Prof-specialty      Husband  White    Male             0   
2          Adm-clerical    Own-child  White  Female             0   
3         Other-service    Own-child  White  Female             0   
4          Adm-clerical         Wife  White  Female             0   
...                 ...          ...    ...     ...           ...   
9995  Machine-op-inspct      Husband  Black    Male          4386   
9996      Other-service    Own-child  White    Male             0   
9997                  ?    Unmarried  Black  Female             0   
9998  Machine-op-inspct      Husband  White    Male             0   
9999  Machine-op-inspct    Unmarried  White    Male             0   

      capital-loss  hours-per-week native-country Income-category  
0                0              40  United-States            >50K  
1                0              40              ?            >50K  
2                0              40  United-States           <=50K  
3                0              35  United-States           <=50K  
4                0              40  United-States           <=50K  
...            ...             ...            ...             ...  
9995             0              40  United-States            >50K  
9996             0              30  United-States           <=50K  
9997             0              15  United-States           <=50K  
9998             0              40  United-States            >50K  
9999             0              40  United-States           <=50K  

[10000 rows x 15 columns]

Question 1

a) We clean and prepare the dataset by evaluating several aspects. We remove the NaN rows as they do not contribute. A possible way of doing this could be by interpolating new values, but because the value set is discrete and individual rather than continuous, it is better to remove them. 

A read-through of the dataset tells us that there are rows with the "?" value. Rows with "?" do not signify any particular value, except the "?" if one was looking for that. Therefore, it is better to remove them. With the reasoning being because there are 10 000 values to consider, the few "?" rows can be removed without there being corrosion of the final result. 

For binary columns, like gender, we can rename the columns to 0 and 1 in order to decrease processing time. This can also be done with race. The object columns were done so accordingly.  

With regards to whether any columns could be removed due to similarities, the education and education-num are interchangeable, and one of them should be removed. But only after a short explanation of what number represents what education level attained. The relationship and marital status columns also seem similar, but after a read-through it is seen that they have varieties and should stay intact. The column called fnlwgt was also removed because it is not understood how the values have been calculated. 

A consideration was done to change the final column of income category with the intention of making it easier to process for the computer. But it was decided to keep it intact. 

In [133]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              10000 non-null  int64 
 1   workclass        10000 non-null  object
 2   fnlwgt           10000 non-null  int64 
 3   education        10000 non-null  object
 4   education-num    10000 non-null  int64 
 5   marital-status   10000 non-null  object
 6   occupation       10000 non-null  object
 7   relationship     10000 non-null  object
 8   race             10000 non-null  object
 9   sex              10000 non-null  object
 10  capital-gain     10000 non-null  int64 
 11  capital-loss     10000 non-null  int64 
 12  hours-per-week   10000 non-null  int64 
 13  native-country   10000 non-null  object
 14  Income-category  10000 non-null  object
dtypes: int64(6), object(9)
memory usage: 1.1+ MB


There are no null rows in any of the columns. This means that no further action is needed in this aspect.

In [134]:
numeric = df.select_dtypes(exclude=['object'])
file_object = df.drop(columns=numeric.columns, axis=1)
file_object

workclass     education      marital-status         occupation  \
0      Private  Some-college  Married-civ-spouse  Machine-op-inspct   
1      Private       Masters  Married-civ-spouse     Prof-specialty   
2      Private       HS-grad            Divorced       Adm-clerical   
3      Private     Assoc-voc       Never-married      Other-service   
4      Private  Some-college  Married-civ-spouse       Adm-clerical   
...        ...           ...                 ...                ...   
9995   Private       HS-grad  Married-civ-spouse  Machine-op-inspct   
9996   Private  Some-college       Never-married      Other-service   
9997         ?     Bachelors  Married-spouse-abs                  ?   
9998   Private       HS-grad  Married-civ-spouse  Machine-op-inspct   
9999   Private       HS-grad       Never-married  Machine-op-inspct   

     relationship   race     sex native-country Income-category  
0         Husband  White    Male  United-States            >50K  
1         Husband  White    Male              ?            >50K  
2       Own-child  White  Female  United-States           <=50K  
3       Own-child  White  Female  United-States           <=50K  
4            Wife  White  Female  United-States           <=50K  
...           ...    ...     ...            ...             ...  
9995      Husband  Black    Male  United-States            >50K  
9996    Own-child  White    Male  United-States           <=50K  
9997    Unmarried  Black  Female  United-States           <=50K  
9998      Husband  White    Male  United-States            >50K  
9999    Unmarried  White    Male  United-States           <=50K  

[10000 rows x 9 columns]

In [135]:
variables = ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']
for variable in variables:
    df[variable] = df[variable].replace('?', np.nan)
df

age workclass  fnlwgt     education  education-num      marital-status  \
0      36   Private  136629  Some-college             10  Married-civ-spouse   
1      56   Private   98361       Masters             14  Married-civ-spouse   
2      32   Private  222548       HS-grad              9            Divorced   
3      24   Private  234640     Assoc-voc             11       Never-married   
4      25   Private   74977  Some-college             10  Married-civ-spouse   
...   ...       ...     ...           ...            ...                 ...   
9995   43   Private  313022       HS-grad              9  Married-civ-spouse   
9996   22   Private  231085  Some-college             10       Never-married   
9997   60       NaN  386261     Bachelors             13  Married-spouse-abs   
9998   47   Private  188081       HS-grad              9  Married-civ-spouse   
9999   25   Private  212304       HS-grad              9       Never-married   

             occupation relationship   race     sex  capital-gain  \
0     Machine-op-inspct      Husband  White    Male             0   
1        Prof-specialty      Husband  White    Male             0   
2          Adm-clerical    Own-child  White  Female             0   
3         Other-service    Own-child  White  Female             0   
4          Adm-clerical         Wife  White  Female             0   
...                 ...          ...    ...     ...           ...   
9995  Machine-op-inspct      Husband  Black    Male          4386   
9996      Other-service    Own-child  White    Male             0   
9997                NaN    Unmarried  Black  Female             0   
9998  Machine-op-inspct      Husband  White    Male             0   
9999  Machine-op-inspct    Unmarried  White    Male             0   

      capital-loss  hours-per-week native-country Income-category  
0                0              40  United-States            >50K  
1                0              40            NaN            >50K  
2                0              40  United-States           <=50K  
3                0              35  United-States           <=50K  
4                0              40  United-States           <=50K  
...            ...             ...            ...             ...  
9995             0              40  United-States            >50K  
9996             0              30  United-States           <=50K  
9997             0              15  United-States           <=50K  
9998             0              40  United-States            >50K  
9999             0              40  United-States           <=50K  

[10000 rows x 15 columns]

In [136]:
df=df.dropna()

We see that '?' has first been transformed to NaN and then dropped. 

In [137]:
from tabulate import tabulate

In [138]:
all_data = [['education-number', 'education-level'],
            [1,'Preschool'], [2, '1st-4th'], [3, '5th-6th' ], [4, '7th-8th'], [5, '9th'], [6, '10th'],
            [7, '11th'], [8, '12th'], [9, 'HS-grad'], [10, 'Some-college'], [11, 'Assoc-voc'], [12, 'Assoc-acdm'],
            [13, 'Bachelors'], [14, 'Masters'], [15, 'Prof-school'], [16, 'Doctorate']]
#table1 = tabulate(all_data)
table2 = tabulate(all_data,headers='firstrow')
 
print(table2)

  education-number  education-level
------------------  -----------------
                 1  Preschool
                 2  1st-4th
                 3  5th-6th
                 4  7th-8th
                 5  9th
                 6  10th
                 7  11th
                 8  12th
                 9  HS-grad
                10  Some-college
                11  Assoc-voc
                12  Assoc-acdm
                13  Bachelors
                14  Masters
                15  Prof-school
                16  Doctorate


In [139]:
df = df.drop('education', axis=1)
df = df.drop('fnlwgt', axis = 1)

b) We are asked to evaluate the performance of three basic classifiers on the dataset. The three classifiers are: a decision tree with at most depth 3, a neural network with at most 10 hiddet networks and 1-NN. 

In [140]:
df.rename(columns={'marital-status':'maritalstatus'}, inplace=True)
df.rename(columns={'native-country':'nativecountry'}, inplace=True)
df.rename(columns={'Income-category':'Incomecategory'}, inplace=True)

In [141]:
df.workclass.unique()

array(['Private', 'Local-gov', 'Self-emp-not-inc', 'Federal-gov',
       'State-gov', 'Self-emp-inc', 'Without-pay'], dtype=object)

In [142]:
df.maritalstatus.unique()

array(['Married-civ-spouse', 'Divorced', 'Never-married', 'Separated',
       'Widowed', 'Married-spouse-abs', 'Married-AF-spouse'], dtype=object)

In [143]:
df.occupation.unique()

array(['Machine-op-inspct', 'Adm-clerical', 'Other-service',
       'Exec-managerial', 'Craft-repair', 'Farming-fishing',
       'Transport-moving', 'Prof-specialty', 'Sales', 'Handlers-cleaners',
       'Tech-support', 'Priv-house-serv', 'Protective-serv',
       'Armed-Forces'], dtype=object)

In [144]:
df.relationship.unique()

array(['Husband', 'Own-child', 'Wife', 'Not-in-family', 'Unmarried',
       'Other-relative'], dtype=object)

In [145]:
df.race.unique()

array(['White', 'Black', 'Asian-Pac-Islander', 'Other',
       'Amer-Indian-Eskimo'], dtype=object)

In [146]:
df.nativecountry.unique()

array(['United-States', 'Puerto-Rico', 'Columbia', 'Italy', 'England',
       'Mexico', 'Poland', 'Philippines', 'Nicaragua', 'Hungary', 'India',
       'Cuba', 'Portugal', 'Germany', 'Canada', 'Guatemala', 'Cambodia',
       'France', 'Dominican-Republic', 'Ecuador', 'El-Salvador', 'Japan',
       'Hong', 'Vietnam', 'Taiwan', 'Haiti', 'China', 'Jamaica', 'South',
       'Laos', 'Trinadad&Tobago', 'Peru', 'Ireland', 'Scotland', 'Iran',
       'Outlying-US-Guam-USVI', 'Yugoslavia', 'Greece', 'Thailand',
       'Honduras', 'Holand-Netherlands'], dtype=object)

In [147]:
c = {'Private':1, 'Self-emp-not-inc':2, 'Self-emp-inc':3, 'Federal-gov':4, 'Local-gov':5, 
     'State-gov':6, 'Without-pay':7, 'Never-worked':8}
df['workclass'] = df['workclass'].map(c)

c = {'Married-civ-spouse':1, 'Divorced':2, 'Never-married':3, 'Separated':4,
       'Widowed':5, 'Married-AF-spouse':6, 'Married-spouse-abs':7}
df['maritalstatus'] = df['maritalstatus'].map(c)

c = {'Machine-op-inspct':1, 'Prof-specialty':2, 'Adm-clerical':3,
       'Other-service':4, 'Exec-managerial':5, 'Craft-repair':6,
       'Farming-fishing':7, 'Priv-house-serv':8, 'Transport-moving':9, 'Sales':10,
       'Handlers-cleaners':11, 'Tech-support':12, 'Protective-serv':13,
       'Armed-Forces':14}
df['occupation'] = df['occupation'].map(c)

c = {'Male':1, 'Female':2}
df['sex'] = df['sex'].map(c)

c = {'Husband':1, 'Own-child':2, 'Wife':3, 'Not-in-family':4, 'Unmarried':5,
       'Other-relative':6}
df['relationship'] = df['relationship'].map(c)

c = {'White':1, 'Black':2, 'Asian-Pac-Islander':3, 'Amer-Indian-Eskimo':4,
       'Other':5}
df['race'] = df['race'].map(c)

c = {'United-States':1, 'Puerto-Rico':2, 'Columbia':3, 'Italy':4,
       'England':5, 'France':6, 'Mexico':7, 'Poland':8, 'Philippines':9,
       'Nicaragua':10, 'Hungary':11, 'India':12, 'Cuba':13, 'Portugal':14, 'Germany':15,
       'Canada':16, 'Guatemala':17, 'Cambodia':18, 'Dominican-Republic':19, 'Ecuador':20,
       'South':21, 'El-Salvador':22, 'Japan':23, 'Hong':24, 'Vietnam':25, 'Taiwan':26,
       'Haiti':27, 'China':28, 'Jamaica':29, 'Laos':30, 'Trinadad&Tobago':31, 'Peru':32,
       'Ireland':33, 'Scotland':34, 'Iran':35, 'Outlying-US-Guam-USVI':36,
       'Yugoslavia':37, 'Greece':38, 'Thailand':39, 'Honduras':40,
       'Holand-Netherlands':41}
df['nativecountry'] = df['nativecountry'].map(c)

In [148]:
from sklearn import tree
from sklearn.model_selection import train_test_split

In [149]:
X = df.drop('Incomecategory', axis=1)
y = df['Incomecategory']

In [150]:
X.head()

age  workclass  education-num  maritalstatus  occupation  relationship  \
0   36          1             10              1           1             1   
2   32          1              9              2           3             2   
3   24          1             11              3           4             2   
4   25          1             10              1           3             3   
5   50          5              9              3           4             4   

   race  sex  capital-gain  capital-loss  hours-per-week  nativecountry  
0     1    1             0             0              40              1  
2     1    2             0             0              40              1  
3     1    2             0             0              35              1  
4     1    2             0             0              40              1  
5     2    2             0             0              40              1

In [151]:
y.head()

0     >50K
2    <=50K
3    <=50K
4    <=50K
5    <=50K
Name: Incomecategory, dtype: object

In [152]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.30,random_state=99)
X_train.head()

age  workclass  education-num  maritalstatus  occupation  relationship  \
2334   33          2              9              1           9             1   
1082   65          1              9              5           3             5   
613    40          1             11              2           4             4   
9417   45          1              9              2           6             5   
1302   18          1              8              3           7             2   

      race  sex  capital-gain  capital-loss  hours-per-week  nativecountry  
2334     1    1             0             0              60              1  
1082     1    2             0             0              40              1  
613      1    2             0             0              32              1  
9417     1    2             0             0              40              1  
1302     1    1             0             0              55              1

In [153]:
from sklearn.tree import DecisionTreeClassifier

In [154]:
tree = DecisionTreeClassifier(max_depth=3)
tree.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=3)

In [155]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
y_pred = tree.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

       <=50K       0.87      0.94      0.90      2096
        >50K       0.75      0.56      0.64       672

    accuracy                           0.85      2768
   macro avg       0.81      0.75      0.77      2768
weighted avg       0.84      0.85      0.84      2768



In [156]:
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))

[[1969  127]
 [ 295  377]]
0.8475433526011561


We then continue on to investigate the Neural Network classifier. 

In [157]:
from sklearn.neural_network import MLPClassifier

In [158]:
def accuracy(confusion_matrix):
   diagonal_sum = confusion_matrix.trace()
   all_elements = confusion_matrix.sum()
   return diagonal_sum / all_elements

In [159]:
classifier = MLPClassifier(hidden_layer_sizes=(10,5,2), max_iter=300,activation = 'relu',solver='adam',random_state=1)
classifier.fit(X_train, y_train)
y_pred2 = classifier.predict(X_test)

In [160]:
cm = confusion_matrix(y_pred2, y_test)
print(cm)
print(accuracy(cm))

[[1950  311]
 [ 146  361]]
0.8348988439306358


In [161]:
print(classification_report(y_test,y_pred2))

              precision    recall  f1-score   support

       <=50K       0.86      0.93      0.90      2096
        >50K       0.71      0.54      0.61       672

    accuracy                           0.83      2768
   macro avg       0.79      0.73      0.75      2768
weighted avg       0.83      0.83      0.83      2768



The last classifier to be investigated was the 1-NN. Which I presume is a version of kNN. 

In [162]:
from sklearn.neighbors import KNeighborsClassifier

In [163]:
knnclassifier=KNeighborsClassifier(n_neighbors = 1)
knnclassifier.fit(X_train,y_train)
knn_pred = knnclassifier.predict(X_test)
Accuracy_of_knn = accuracy_score(y_test, knn_pred)
cm2 = confusion_matrix(knn_pred, y_test)
print(cm2)
print(Accuracy_of_knn)

[[1839  269]
 [ 257  403]]
0.809971098265896


In [164]:
print(classification_report(y_test,knn_pred))

              precision    recall  f1-score   support

       <=50K       0.87      0.88      0.87      2096
        >50K       0.61      0.60      0.61       672

    accuracy                           0.81      2768
   macro avg       0.74      0.74      0.74      2768
weighted avg       0.81      0.81      0.81      2768



Our evaluation shows that the Decision tree classifier has the highest accuracy, with neural network coming in second and knn classifier coming in last. 

c) We are asked to apply the bagging method using the three classifiers introduced in the previous question. Furthermore, we are also to investigate the performance of each of these classifiers as the ensemble size increases from 2 to 20. 

In [165]:
from sklearn.ensemble import BaggingClassifier

In [166]:
#We create a base classifier for the decision tree method

DTC = DecisionTreeClassifier()
for e in range(2,20,2):

    model = BaggingClassifier(estimator=DTC, n_estimators=e, random_state=0, max_samples=0.5, max_features=1.0, bootstrap=True)
    classifiers = model.fit(X_train, y_train)
    y_pred3 = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)


0.8179190751445087
0.8302023121387283
0.8327312138728323
0.8385115606936416
0.8424855491329479
0.8442919075144508
0.8482658959537572
0.8504335260115607
0.8507947976878613


In [167]:
for e in range(2,20,2):
    model = BaggingClassifier(estimator=knnclassifier, n_estimators=e, random_state=0, max_samples=0.5, max_features=1.0, bootstrap=True)
    classifiers = model.fit(X_train, y_train)

    y_pred3 = model.predict(X_test)
 
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)

0.8208092485549133
0.821893063583815
0.826228323699422
0.8283959537572254
0.8302023121387283
0.8305635838150289
0.8305635838150289
0.8302023121387283
0.8323699421965318


In [168]:
for e in range(2,20,2):
    model = BaggingClassifier(estimator=classifier, n_estimators=e, random_state=0, max_samples=0.5, max_features=1.0, bootstrap=True)
    classifiers = model.fit(X_train, y_train)

    y_pred3 = model.predict(X_test)
 
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)

0.8182803468208093
0.7835982658959537
0.7929913294797688
0.7991329479768786


C:\Users\Furka\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


0.798771676300578


C:\Users\Furka\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7958815028901735


C:\Users\Furka\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7969653179190751


C:\Users\Furka\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


0.798771676300578


C:\Users\Furka\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7814306358381503


The latter part of the question asks us by using the best performing ensemble size, to 
investigate how changing the number of instances in the bootstrap samples affects
classification performance. We see that for the Decision tree classifier the accuracy increases as more estimators are introduced. This is sort of true for the 1-NN classifier. If we only measured the first 2 and the last 20 estimators, then we could conclude an improvement in accuracy, albeit a smaller than increase than for the Decision tree classifier. However, further inspection shows that the values for each 2 estimator increase do not necessary follow a trajection. It varies whether it increases or decreased whenever 2 estimators are introduced. With the MLP classifier we see that the accuracy varies for each estimator introduced as well. But if measured at the first two estimators and last 20, we see that there has been a decrease in accuracy from 2 to 20 estimators.

For classifiers where there is no clear tendency with improved accuracy to observe: this could be due to samples being drawn with replacement. 

The ensemble size shown in the examples above is the best according to my computer and know-how. 

d) In this question, we are meant to the same task as in c), but we exchange the bagging method with the random subspace method. 

In [169]:

DTC = DecisionTreeClassifier()
for e in range(2,20,2):

    model = BaggingClassifier(estimator=DTC, n_estimators=e, random_state=0, max_samples=1.0, max_features=0.5, bootstrap=False)
    classifiers = model.fit(X_train, y_train)
    y_pred3 = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)


0.8363439306358381
0.8385115606936416
0.8439306358381503
0.8529624277456648
0.8569364161849711
0.8601878612716763
0.861271676300578
0.8627167630057804
0.8601878612716763


In [170]:
for e in range(2,20,2):
    model = BaggingClassifier(estimator=knnclassifier, n_estimators=e, random_state=0, max_samples=1.0, max_features=0.5, bootstrap=False)
    classifiers = model.fit(X_train, y_train)

    y_pred3 = model.predict(X_test)
 
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)

0.8106936416184971
0.815028901734104
0.8240606936416185
0.8316473988439307
0.8363439306358381
0.8410404624277457
0.8424855491329479
0.8435693641618497
0.8460982658959537


In [171]:
for e in range(2,20,2):
    model = BaggingClassifier(estimator=classifier, n_estimators=e, random_state=0, max_samples=1.0, max_features=0.5, bootstrap=False)
    classifiers = model.fit(X_train, y_train)

    y_pred3 = model.predict(X_test)
 
    accuracy = accuracy_score(y_test, y_pred3)
    print(accuracy)

0.7915462427745664
0.8005780346820809
0.8023843930635838
0.7973265895953757
0.791907514450867
0.7976878612716763
0.7958815028901735
0.7994942196531792
0.7940751445086706


First of all it should be emphasized that the best ensemble size is the one suggested in the examples above. Higher or lower max_features decreases the accuracy. 

We see yet again that the Decision tree classifier increases the accuracy as the number of estimators increases, this tendency continues all the way up to 20 estimators. This tendency is also seen for 1-NN classifiers, although the accuracy is slightly lower for this classifier compared to the Decision tree classifier. The MLP classifier is almost unchanged as the number of estimators increases. This could due to the attempt to limit the effect of samples drawn with replacement. 

e) Based on the lectures, we expect the Decision tree classifier and neural networks to benefit from the bagging method. While random subspace should work well for k-NN. It must here be emphasized that the 1-NN is k-NN with one nearest neighbour. And a k-NN model fits a specific point in the data with the N nearest data points in your training set. For 1-NN this point depends only of 1 single other point

When we consider the ensemble strategy for the classifiers we used, we have to take into consideration the size of the database, the number of estimators used, whether we should use bagging method or random subspace or neither of them, the number of features and the relation between object and integers (this aspect was included due to the work in this assignment had some processing time issues). The strategy for the Decision tree classifier must include that the classifier is the simplest classifier of those evaluated. It gives high accuracy and by increasing the number of estimators, the accuracy increases. The number of estimators should be 20 (highest allowed) and the bagging method should be used as it had higher accuracy then both Decision tree classifier alone and random subspace technique used. 

For the neural network method used, we see that when neither bagging nor random subspace methods are included, then the accuracy is at it's highest. This suggests that the best ensemble strategy to use. It should be added that Neural networks need large training data compared to KNN to achieve sufficient accuracy. Neural networks needs lot of hyperparameter tuning compared to KNN. The database used in this project should perhaps be bigger to optimize Neural networks. 

Both bagging and random subspace improve the 1-NN classifier. And the bagging method improves it more than the random subspace. This is in line with the lectures. Furthermore, the ensemble strategy should include to have 20 estimators.

Ensemble diversity refers to the idea that the individual models prediction in an ensemble should be as different from each other as possible. Different models make different type of errors. In the evaulation above, we tried several methods to improve diversity. Bagging and random subspace methods are fairly obvious. Likewise, the variation of max features and samples. It was seen that 0.5 in max features was optimal. To understand this, we introduce the terms good and bad diversity. Bad diversity here is that we increase the amount of features, but get worse accuracy. Thereby suggesting that we include diversity just for the sake of it. Concluding that the models are diverse enough to obtain necessary accuracy. It should be added that increasing estimators increases accuracy for two of the classifiers.

On what it would take to improve performance further, it is clear that a large database would improve the neural network classifier. While perhaps more estimators could give better accuracy (this was not attempted in this work as it was specified 20 estimators was the upper limit). Besides this, the voting ensemble where one uses several classifiers could be tried out. We could also atttempt different algorithms to capture different patterns in the data. 

Question 2

a) In a) we are asked to define the features to use in the state representation of your RL agent, the reward function for the same RL agent and the length of an episode. We are meant to distribute the balls so that ideally each ball is in each bin. This means that one feature should be what bin a ball is in. Another feature should be how man balls are in each bin and there should also be a feature that verifies whether when one moves a ball to B, then if B-1 > 0 and B+1 < 9. 

The second aspect to consider was the concept of the reward function. A reward function is a crucial component of reinforcement learning, a branch of artificial intelligence (AI) that enables agents to learn from their own actions and feedback. We therefore suggest a reward function where each ball in each bin is optimal reward and the initial position of all balls in one bin is zero reward. 

The last point was the length of the episodes. This should be 10, just like the number of balls and bins. 

In [195]:
pip install gymnasium

Note: you may need to restart the kernel to use updated packages.


In [196]:
pip install shimmy

Note: you may need to restart the kernel to use updated packages.


In [279]:
pip install stable-baselines3[extra]

Note: you may need to restart the kernel to use updated packages.


In [280]:
import gymnasium as gym
from gym import spaces
import stable_baselines3 as sb3
import torch as th
import shimmy
import random 

from stable_baselines3.common.env_checker import check_env

b) Second part of the second question wants us to implement how the agent interacts with the environment (i.e., the step function, restart function, init function etc.)

We will explain the three functions above before making the code to implement them. 
The step() funcitonreturns the current observations and the one-step reward
The reset() function sets the environment to an initial state and computes the initial value of the observation.
The init() is executed when the class is being initialized. 

In [285]:
import gym
from gym import spaces

class CustomEnv(gym.Env):
    #We will begin with init function. We include the number of balls, bins (start position) and number of actions (-1, 0, 1)
    #symbolizing whether place ball right, same or left.
    def __init__(self):
        self.bins = [5] * 10
        self.balls = list(range(10))
        self.actions = [-1, 0, 1]
        self.observation_space = spaces.Box(low=0, high=1, shape=(10, 10), dtype=np.int32)
        self.action_space = spaces.Discrete(len(self.actions))
            
    def reset(self):
        self.bins = [5] * 10
        self.balls = list(range(10))
        return self._get_state()
    
    def step(self, action):
        ball = self.balls.pop(0)
        binpos = ball
        if action == 0:
            pass
        elif action == 1 and binpos < 9:
            binpos += 1
        elif action == 2 and binpos > 0:
            binpos -= 1
        self.bins[ball] = binpos
        reward = self._get_reward()
        done = len(self.balls) == 0
        return self._get_state(), reward, done, {}
    
    def _get_state(self):
        state = np.zeros((10, 10), dtype=np.int32)
        for i, ball in enumerate(self.balls):
            state[ball, self.bins[ball]] = 1
        return state
    
    def _get_reward(self):
        return -np.sum(np.abs(np.array(self.bins) - np.mean(self.bins)))
    
#We continue on to train the RL agent.
env = CustomEnv()
C = np.zeros((100000000, env.action_space.n))
a = 0.001
g = 0.9
for i in range(10):
    state = env.reset()
    action = np.argmax(C[i])
    if random.random() < 0.001:
        action = random.choice(range(env.action_space.n))
    done = False
    while not done:
        new_state, reward, done, _ = env.step(action)
        new_action = np.argmax(C[i])
        if random.random() < 0.1:
            new_action = random.choice(range(env.action_space.n))
        C[i+1, action] += a * (reward + g * np.max(C[i, :]) - C[i, action])

        state = new_state
        action = new_action

print(env.bins)


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [286]:
print(env.observation_space)

Box(0, 1, (10, 10), int32)


In [287]:
def check_env(env, warn=True):
    if not isinstance(env, gym.Env):
        if warn:
            print("Warning: env is not an instance of gym.Env")
        return False

    if not isinstance(env.observation_space, spaces.Box):
        if warn:
            print("Warning: observation space is not an instance of spaces.Box")
        return False

    if not isinstance(env.action_space, spaces.Discrete):
        if warn:
            print("Warning: action space is not an instance of spaces.Discrete")
        return False

    return True

In [288]:
env = CustomEnv()
check_env(env)

True

c) This part of the question asked for the following: Compare the RL strategies PPO, DQN and A2C for this problem. Are they able to learn a consistent winning policy after (i) 50,000 episodes, (ii) 100,000 episodes and (iii) 200,000
episodes? We will first import them and try out the different RL strategies separately to compare them. OpenAI defines PPO as an on-policy algorithm. PPO can be used for environments with either discrete or continuous action spaces. This is a contrast to DQN and A2C that are both value-based approaches that optimize a value function.



In [293]:
from stable_baselines3 import DQN, PPO, A2C
from typing import Callable

In [294]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=dict(pi=[32, 32, 32], vf=[32, 32, 32]))

# Train the agent (if you omit the policy_kwargs argument, default architecture will be used)
model = A2C('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 10       |
|    ep_rew_mean        | -169     |
| time/                 |          |
|    fps                | 739      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -1.08    |
|    explained_variance | 0.462    |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | -36.2    |
|    value_loss         | 1.24e+03 |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 10       |
|    ep_rew_mean        | -168     |
| time/                 |          |
|    fps                | 737      |
|    iterations         | 200      |
|    time_elapsed 

In [298]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[32,33])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -167     |
| time/              |          |
|    fps             | 1881     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 10           |
|    ep_rew_mean          | -166         |
| time/                   |              |
|    fps                  | 1280         |
|    iterations           | 2            |
|    time_elapsed         | 3            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0010694516 |
|    clip_fraction        | 0            |
|    clip_range           | 0.2          |
|    en

In [297]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[32,33])
model = DQN('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 10       |
|    ep_rew_mean      | -167     |
|    exploration_rate | 0.992    |
| time/               |          |
|    episodes         | 4        |
|    fps              | 4277     |
|    time_elapsed     | 0        |
|    total_timesteps  | 40       |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 10       |
|    ep_rew_mean      | -166     |
|    exploration_rate | 0.985    |
| time/               |          |
|    episodes         | 8        |
|    fps              | 6204     |
|    time_elapsed     | 0        |
|    total_timesteps  | 80       |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 10       |
|    ep_rew_mean      | -164   

We have here only included the code for 50 000 time steps. We can use the ep_rew_mean to partially evaluate how well the RL agents do during the training. The higher the value, the better it has performed. After 50 000 the PPO has an ep_rew_mean of -117, while A2C has an ep_rew_mean of -121 and DQN has one of -169. OpenAI defines PPO as an on-policy algorithm. PPO can be used for environments with either discrete or continuous action spaces. This is a contrast to DQN and A2C that are both value-based approaches that optimize a value function. All three models are able to learn, but both and theory show that PPO learns quickest due to its ability to directly optimize policy.PPO achieves a high reward after 50 000, and is the best performing agent. 

d) PPO was used for the task of neural networks. We use the same code as before, but specify that the net_arch parameter has 3 hidden layers with 64 neurons neurons each, 3 hidden layers with 16 neurons each, 2 hidden layers with 16 neurons and finally a single hidden layer with 32, 16, 4 and 2 neurons.  

In [299]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[64,64,64])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -169     |
| time/              |          |
|    fps             | 1916     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 10           |
|    ep_rew_mean          | -169         |
| time/                   |              |
|    fps                  | 1120         |
|    iterations           | 2            |
|    time_elapsed         | 3            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0009601611 |
|    clip_fraction        | 0            |
|    clip_range           | 0.2          |
|    en

In [300]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[16,16,16])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -170     |
| time/              |          |
|    fps             | 1743     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 10          |
|    ep_rew_mean          | -167        |
| time/                   |             |
|    fps                  | 1216        |
|    iterations           | 2           |
|    time_elapsed         | 3           |
|    total_timesteps      | 4096        |
| train/                  |             |
|    approx_kl            | 0.001587895 |
|    clip_fraction        | 0           |
|    clip_range           | 0.2         |
|    entropy_loss   

In [301]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[16,16])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -169     |
| time/              |          |
|    fps             | 1994     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 10           |
|    ep_rew_mean          | -167         |
| time/                   |              |
|    fps                  | 1384         |
|    iterations           | 2            |
|    time_elapsed         | 2            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0043947035 |
|    clip_fraction        | 0.00083      |
|    clip_range           | 0.2          |
|    en

In [302]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[32])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -170     |
| time/              |          |
|    fps             | 2093     |
|    iterations      | 1        |
|    time_elapsed    | 0        |
|    total_timesteps | 2048     |
---------------------------------
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 10          |
|    ep_rew_mean          | -164        |
| time/                   |             |
|    fps                  | 1509        |
|    iterations           | 2           |
|    time_elapsed         | 2           |
|    total_timesteps      | 4096        |
| train/                  |             |
|    approx_kl            | 0.015341832 |
|    clip_fraction        | 0.0971      |
|    clip_range           | 0.2         |
|    entropy_loss   

In [303]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[16])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -170     |
| time/              |          |
|    fps             | 1975     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 10           |
|    ep_rew_mean          | -165         |
| time/                   |              |
|    fps                  | 1484         |
|    iterations           | 2            |
|    time_elapsed         | 2            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0054580197 |
|    clip_fraction        | 0.00801      |
|    clip_range           | 0.2          |
|    en

In [306]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[4])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -168     |
| time/              |          |
|    fps             | 1956     |
|    iterations      | 1        |
|    time_elapsed    | 1        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 10           |
|    ep_rew_mean          | -169         |
| time/                   |              |
|    fps                  | 1498         |
|    iterations           | 2            |
|    time_elapsed         | 2            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0014899324 |
|    clip_fraction        | 0            |
|    clip_range           | 0.2          |
|    en

In [304]:
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[2])
model = PPO('MlpPolicy', env, policy_kwargs=policy_kwargs, verbose=1).learn(50000)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 10       |
|    ep_rew_mean     | -172     |
| time/              |          |
|    fps             | 2181     |
|    iterations      | 1        |
|    time_elapsed    | 0        |
|    total_timesteps | 2048     |
---------------------------------
-------------------------------------------
| rollout/                |               |
|    ep_len_mean          | 10            |
|    ep_rew_mean          | -169          |
| time/                   |               |
|    fps                  | 1589          |
|    iterations           | 2             |
|    time_elapsed         | 2             |
|    total_timesteps      | 4096          |
| train/                  |               |
|    approx_kl            | 0.00023115595 |
|    clip_fraction        | 0             |
|    clip_range           | 0.2       

A single hidden layer of 16 neurons is the minimal architechture that gives us a winning policy. The example of 2 gives too few layer to learn. I deem that the value loss of 115 is too high and indicates that the agent is not learning effectively.